hw_2_functions.py

```diff
@@ -0,0 +1,284 @@
+
+
+# **4)Given a 3x3 matrix of a completed tic-tac-toe game, create a function that returns whether the game is a win
+# for "X", "O", or a "Draw", where "X" and "O" represent themselves on the matrix, and "E" represents an empty spot.
+# Example:
+# tic_tac_toe([
+#     ["X", "O", "X"],
+#     ["O", "X", "O"],
+#     ["O", "X", "X"]
+# ]) ➞ "X"
+#
+# tic_tac_toe([
+#     ["O", "O", "O"],
+#     ["O", "X", "X"],
+#     ["E", "X", "X"]
+# ]) ➞ "O"
+#
+# tic_tac_toe([
+#     ["X", "X", "O"],
+#     ["O", "O", "X"],
+#     ["X", "X", "O"]
+# ]) ➞ "Draw"
+
+
+# Run tic-tac-toe game matrix analysis
+def tic_tac_toe_analyze(game_matrix):
+    """Function accepts matrix for finished tic-tac-toe game - a list of 3 lists each containing 3 chars:
+    'X', 'O', 'E' - crosses, zeros or empty cells
+    based on
+    """
+    results = fill_x_o_positions(game_matrix)
+    win = check_win(results)
+    print_matrix(game_matrix)
+    if win in ['X', 'O']:
+        print(f"\n{win} wins in this game\n")
+    elif win == 'D':
+        print('\nThis is a draw game\n')
+    elif win == 'U':
+        print("\nStill more than 1 move to make. Might be a draw game though..\n")
+    elif win == 'F':
+        print("\nSomething is wrong with this game. Looks like some O's are missing\n")
+    print("-"*20)
+
+
+# Check if one of players wins
+def check_win(filled_matrix):
+    # All possible winning lines
+    winner_lines = [[1, 2, 3], [4, 5, 6], [7, 8, 9], [1, 4, 7], [2, 5, 8], [3, 6, 9], [1, 5, 9], [3, 5, 7]]
+    global winner
+    winner = None
+    x_pos = filled_matrix[0]
+    o_pos = filled_matrix[1]
+    e_pos = filled_matrix[2]
+    if len(x_pos) - len(o_pos) > 1:
+        winner = 'F'
+        return winner
+    # If one move is left and number of X's is the same as O's then we add last cell to X
+    if len(e_pos) == 1 and len(x_pos) == len(o_pos):
+        x_pos.append(e_pos[0])
+    # Loop to check if any of the winning lines is in the list of X/O positions
+    for line in winner_lines:
+        if all(elem in x_pos for elem in line):
+            winner = 'X'
+            break
+        elif all(elem in o_pos for elem in line):
+            winner = 'O'
+            break
+    if winner not in ['X', 'O']:
+        # analysis for more than 1 empty cell is out of scope of this task
+        if len(e_pos) >= 2:
+            winner = 'U'
+        else:
+            winner = 'D'
+    return winner
+
+
+def fill_x_o_positions(matrix):
+    position = 0
+    x_positions, o_positions, e_positions = [], [], []
+    for line in matrix:
+        for item in line:
+            position += 1
+            if item == 'X':
+                x_positions.append(position)
+            elif item == 'O':
+                o_positions.append(position)
+            else:
+                e_positions.append(position)
+    return x_positions, o_positions, e_positions
+
+
+def print_matrix(matrix):
+    my_printer = pprint.PrettyPrinter(width=30)
+    my_printer.pprint(matrix)
+
```

Comment:
I'd like to share some very short solution with you:

def tic_tac_toe(matrix):
    result = [''.join(matrix[i][j] for j in range(3)) for i in range(3)] + \
             [''.join(matrix[j][i] for j in range(3)) for i in range(3)] + \
             [''.join(matrix[i][i] for i in range(3))] + \
             [''.join(matrix[i][2-i] for i in range(3))]
    return 'X' if 'XXX' in result else 'O' if 'OOO' in result else 'Draw'